README.md
# linux-shell
>**simple linux shell with C**

program that represents a linux shell and supports linux shell commands
i have added some additional commands which they focus on file reading things

list of additional commands:
1. a -> returns the 1st part of given txt file
2. b-> returns the most frequent string in given txt file
3. c-> removes all empty spaces such as: space, \n, \t
4. d-> shows the not commented lines in the txt file
5. f-> returns number of lines in the txt file
6. g-> returns the first n line of the txt file

it has a simple history txt file which contains all program recent commands

some examples of commands:
![](welcome.png)
![](directory.png)
![](linux-command.png)
![](additional.png)
![](exit.png)


history.txt
exit
ls
is
pwd
ls
 lls
exit
ls
a a.txt
b b.txt
exit
b b.txt
b b.txt
exit
b b.txt
b b,txt
b b.txt
b b.txt
b b.txt
exit
b b.txt
b b.txt
exit
b b.txt
exit
b b.txt
exit
b b.txt
b b.txt
exit
b b.txt
exit
b b.txt
b b.txt
b b.txt
exit
b b.txt
exit
b b.txt
b b.txt
b b.txt
exit
b b.txt
exit
b b.txt
a sds
b sds
b sdlksfm
exit
b b.txt
a
a
a a.txt
a
exit
b
a
exit
exit
exit
exit
b bs
exit
b b
b kks
a dds
exit
exit
exit
exit
a sd
a dss
a dsds
exit
exit
exit
exit
b dssd
oos
exit
b lsdss
b ds
exit
b dsds
b ldslds
b dskdsd
b b.txt
a dsds
a dsldfffs
a fdkdfl
a a.txt
exit
b b.txt
b dsds
exit
b b;lkds
a ldksklds
a ldss
b l;sa
exit
b lds;
b dksldlsds
a kdlsdk
b dlkskds
exit
b dsd
b ds;ds
b b.txt
a ;sdfkksd
a a.txt
c
exit
c ldsds
c c.txt
exit
c c.txt
exit
c 
c ks
c c.txt
c c.txt
exit
c c.txt
exit
c c.txt
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exi
exit
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
ls
a
a ldss
a a.txt
b b.txt
c c.txt
ls
access
accton
alias
awk
bzmore
bzmore
ls
exitt
exit
ls
ff
q
s
a
a
a test.txt
a a.txt.
a a.txt
b b.txt
c c.txt
exit
c c.txt
exit
exit
c c.txt
exit
c c.txt
exit
c c.txt
C:\Users\mohammad\Documents\GitHub\linux-shell\c.txt
exit
c c.txt
exit
c c.txt
c c.txt

c c.txt
c c.txt
exit
c c.txt
exit
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
c c.txt
exit
c c.txt
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
c c.txt
exit
d d.txt
exit
d d.txt
exit
d d.txt
d d.txt
exit
d d.txt
exit
d d.txt
exit
d d.txt
exit
d d.txt
exit
d d.txt
exit
d d.txt
exit
d d.txt
exit
d d.txt
exit
d d.txt
exit
ls
d d.txt
d d.txt
exit
f f.txt
f f.txt
f
f f.txt
f f.txt
f f.txt
f f.txt
exit
f f.txt
exit
g g.txt
exit
g g.txt
c c.txt
exit
c c.txt
pwd
pwd --help
exit
ls
ls
a

lsd
ld
i
a
a 

exit
pwd
exit
sasa
sa
help
exit
cd desktop
cd /c
cd
pwd
pwd --l
cd
exit
cd
cd
pwd
exit
ssds
exit
sa
a
exit
ls
a a.txt
exit
exit
exit
exit
exit
ls
a a.txt
b b.txt
exit
dsada
exit


tests/test.h
//
// Created by mohammad on 12/26/2022.
//

#ifndef LINUX_SHELL_TEST_H
#define LINUX_SHELL_TEST_H


class test {

};


#endif //LINUX_SHELL_TEST_H


tests/test2.c
//#include<stdio.h>
//#include<string.h>
//#include<stdlib.h>
//#include<unistd.h>
//#include<sys/types.h>
//#include<sys/wait.h>
//#include<readline/readline.h>
//#include<readline/history.h>
//#include <sys/wait.h>
//#include <stdbool.h>
//#include <string.h>
//#include <signal.h>
//
//#define clear() printf("\033[H\033[J")
//
//const int MAX = 1000;
//char** TOKEN;
//
//void welcome();
//void printDir();
//void execute(char *input);
//void getInput();
//bool myCommands(char* input);
//void error();
//void a();
//void b();
//void c();
//void d();
//void f();
//void g();
//char** splitter(const char * str, const char * delim);
//bool fileValidator(char* input);
//char* fileToText(FILE* file);
//void addHistory(char* input);
//void handleSigint(int sig);
//int  isEmpty(const char *str);
//FILE* removeEmptyLines(FILE *file);
//char* removeSpaces(char *string);
//void printFile(FILE *fptr);
//
//void splitter2(char* text){
//    // Extract the first token
//    char* temp[MAX];
//    int count =0;
//
//    char * token = strtok(text, " ");
//    // loop through the string to extract all other tokens
//    while( token != NULL ) {
////        printf( " %s\n", token ); //printing each token
//        temp[count] = token;
//        count++;
//        token = strtok(NULL, " ");
//    }
//
//
//
//}
//
//int main() {
//    signal(SIGINT, handleSigint);
//    welcome();
//    getInput();
//    printDir();
//
//    return 0;
//}
//
//char ** splitter(const char * str, const char* delim){
//    /* count words */
//    char * s = strdup(str);
//
//    if (strtok(s, delim) == 0)
//        /* no word */
//        return NULL;
//
//    int nw = 1;
//
//    while (strtok(NULL, delim) != 0)
//        nw += 1;
//
//    strcpy(s, str); /* restore initial string modified by strtok */
//
//    /* split */
//    char ** v = malloc((nw + 1) * sizeof(char *));
//    int i;
//
//    v[0] = strdup(strtok(s, delim));
//
//    for (i = 1; i != nw; ++i)
//        v[i] = strdup(strtok(NULL, delim));
//
//    v[i] = NULL; /* end mark */
//
//    free(s);
//
//    return v;
//}
//
//void getInput() {
//    char* input;
//    printDir();
//    input = readline("\n>>: ");
//    addHistory(input);
//    TOKEN = splitter(input," ");
//
//
//
//    char* temp[MAX];
//    int count =0;
//
//    char * token = strtok(input, " ");
//    // loop through the string to extract all other tokens
//    while( token != NULL ) {
////        printf( " %s\n", token ); //printing each token
//        temp[count] = token;
//        count++;
//        token = strtok(NULL, " ");
//    }
//
//
//    execute(TOKEN[0]);
//}
//
//void execute(char *input) {
//
//    char* argument_list[] = {"", "-l", NULL};
//    argument_list[0] = input;
//    if(!strcmp("exit",input)){
//        printf("\nbye\n");
//        exit(0);
//    }
//
//    else{
//        pid_t child = fork();
//
//        if(child < 0) {
//            printf("error\n");
//            exit(0);
//        }
//
//        else if(child == 0) {
//            if(execvp(input,argument_list) == -1){
//                if(!myCommands(input)){
//                    error();
//                }
//            }
//            exit(0);
//        }
//        else {
//            wait(NULL);
//            getInput();
//        }
//
//    }
//
//}
//
//void welcome(){
//    clear();
//    printf("                  ****************** welcome to my linux shell ******************                   \n");
//    fflush(stdout);
//    sleep(1);
//    clear();
//}
//
//void printDir(){
//    char dir[MAX];
//    char * username;
//    getcwd(dir, sizeof(dir));
//    username = getenv("USER");
//    printf("%s >> %s\n",username,dir);
//}
//
//void error(){
//    printf("wrong command\ntry using help\n");
//    fflush(stdout);
//    sleep(1);
//    clear();
//}
//
//bool myCommands(char* input){
//    if(!strcmp(input,"exit"))
//        exit(0);
//    else if(!strcmp(input,"a")) {
//        a();
//        return true;
//    }
//    else if(!strcmp(input,"b")) {
//        b();
//        return true;
//    }
//    else if(!strcmp(input,"c")) {
//        c();
//        return true;
//    }
//    else if(!strcmp(input,"d")) {
//        d();
//        return true;
//    }
//    else if(!strcmp(input,"f")) {
//        f();
//        return true;
//    }
//    else if(!strcmp(input,"g")) {
//        g();
//        return true;
//    }
//
//    return false;
//}
//
//void g() {
//    if(fileValidator(TOKEN[1])){
//
//        FILE* file;
//        file = fopen(TOKEN[1],"r");
//        char* text = fileToText(file);
//        fclose(file);
//        int lineNumberToShow = 10;
//        int counter = 1;
//
//        for (char c = *text; c; c=*++text) {
//            printf("%c",c);
//            if(c == '\n')
//                counter++;
//            if(counter == lineNumberToShow+1)
//                break;
//        }
//
//    }
//
//    else {
//        printf("no such file\n");
//        error();
//    }
//
//    printf("********* g command *********\n");
//}
//
//void f() {
//    if(fileValidator(TOKEN[1])){
//        FILE* file;
//        int count = 0; // Line counter (result)
//        file = fopen(TOKEN[1], "r");
//
//
//        char* text = fileToText(file);
//        fclose(file);
//
//        for (char c = *text; c; c=*++text) {
//            if(c == '\n')
//                count++;
//        }
//
//        printf("%d",count+1);
//    }
//
//    else {
//        printf("no such file\n");
//        error();
//    }
//
//    printf("********* f command *********\n");
//}
//
//void d() {
//    if(fileValidator(TOKEN[1])){
//        FILE* file;
//        file = fopen(TOKEN[1],"r");
//        char* text = fileToText(file);
//
//        int i=0;
//        char lastchar;
//        for (char c = *text; c; c=*++text) {
//            int hashtag = 35;
//
//            if((c == hashtag) && (lastchar == '\n')){
//                while(c!=0) {
//                    c = *++text;
//                    lastchar = c;
//                }
//                c=*++text;
//            }
//            i++;
//            printf("%c",c);
//            lastchar = c;
//        }
//        fclose(file);
//    }
//
//    else {
//        printf("no such file\n");
//        error();
//    }
//
//    printf("********* c command *********\n");
//}
//
//void c() {
//    if(fileValidator(TOKEN[1])){
//
//        FILE* file;
//        file  = fopen(TOKEN[1], "r");
//
//        printf("\n%s\n",removeSpaces(fileToText(removeEmptyLines(file))));
//
//        fclose(file);
//    }
//
//    else {
//        printf("no such file\n");
//        error();
//    }
//    printf("********* c command *********\n");}
//
//void b() {
//    if(fileValidator(TOKEN[1])){
//        FILE* file;
//        file = fopen(TOKEN[1],"r");
//        char* text = fileToText(file);
//        fclose(file);
//
//        char result;
//        int i, len;
//        int max = -1;
//        int freq[MAX];
//        len = strlen(text);
//
//        for(i = 0; i < len; i++){
//            freq[text[i]]++;
//        }
//        for(i = 0; i < len; i++){
//            if(max < freq[text[i]]){
//                max = freq[text[i]];
//                result = text[i];
//            }
//        }
//
//        if(strcmp(&result," ")) {
//            if(strcmp(&result,"\n"))
//                printf("\\n \n");
//            else
//                printf("space\n");
//        }
//        else {
//            printf("%c\n", result);
//        }
//    }
//
//    else {
//        printf("no such file\n");
//        error();
//    }
//
//    printf("********* b command *********\n");
//}
//
//void a() {
//    if(fileValidator(TOKEN[1])){
//        FILE* file;
//        file = fopen(TOKEN[1],"r");
//        char* text = fileToText(file);
//        fclose(file);
//        char** temp = splitter(text," ");
//        puts(temp[0]);
//    }
//
//    else {
//        printf("no such file\n");
//        error();
//    }
//    printf("********* a command *********\n");
//}
//
//bool fileValidator(char* input){
//    FILE* ptr;
//    ptr = fopen(input, "r");
//    if (NULL == ptr) {
//        return false;
//    }
//    else
//        return true;
//}
//
//char* fileToText(FILE* file){
//    char ch;
//    char text[MAX];
//    int counter=0;
//
//    while (!feof(file)) {
//        ch = fgetc(file);
//        text[counter] = ch;
//        counter++;
//    }
//    fclose(file);
//
//    char* temp = malloc(MAX);
//
//    memcpy(temp,text, sizeof(text));
//    return temp;
//}
//
//void addHistory(char* input){
//    FILE* file;
//    if(fileValidator("history.txt")){
//        file = fopen("history.txt","a");
//        fprintf(file,input);
//        fprintf(file,"\n");
//        fclose(file);
//    }
//    else{
//        file = fopen("history.txt","w");
//        fprintf(file,input);
//        fprintf(file,"\n");
//        fclose(file);
//    }
//}
//
//void handleSigint(int sig){
//    clear();
//    printDir();
//}
//
//void printFile(FILE *fptr)
//{
//    char ch;
//
//    while((ch = fgetc(fptr)) != EOF)
//        putchar(ch);
//}
//
//int isEmpty(const char *str)
//{
//    char ch;
//
//    do
//    {
//        ch = *(str++);
//
//        // Check non whitespace character
//        if(ch != ' ' && ch != '\t' && ch != '\n' && ch != '\r' && ch != '\0')
//            return 0;
//
//    } while (ch != '\0');
//
//    return 1;
//}
//
//FILE* removeEmptyLines(FILE *file){
//    FILE* temp;
//    temp = fopen("remove-blanks.tmp", "w");
//
//
//    /* Exit if file not opened successfully */
//    if (file == NULL || temp == NULL)
//    {
//        printf("Unable to open file.\n");
//        printf("Please check you have read/write previleges.\n");
//
//        exit(EXIT_FAILURE);
//    }
//
//    // Move src file pointer to beginning
//    rewind(file);
//
//    // Remove empty lines from file.
//    char buffer[MAX];
//
//    while ((fgets(buffer, MAX, file)) != NULL)
//    {
//        /* If current line is not empty then write to temporary file */
//        if(!isEmpty(buffer))
//            fputs(buffer, temp);
//    }
//
//    /* Close all open files */
//    fclose(temp);
//    temp = fopen("remove-blanks.tmp", "r");
//    return temp;
//
//}
//
//char * removeSpaces(char *string)
//{
//    // non_space_count to keep the frequency of non space characters
//    int non_space_count = 0;
//
//    //Traverse a string and if it is non space character then, place it at index non_space_count
//    for (int i = 0; string[i] != '\0'; i++)
//    {
//        if (string[i] != ' ')
//        {
//            string[non_space_count] = string[i];
//            non_space_count++;//non_space_count incremented
//        }
//    }
//
//    //Finally placing final character at the string end
//    string[non_space_count] = '\0';
//    return string;
//}
//
//


tests/test.cpp
//#include<stdio.h>
//#include<string.h>
//#include<stdlib.h>
//#include<unistd.h>
//#include<sys/types.h>
//#include<sys/wait.h>
//#include<readline/readline.h>
//#include<readline/history.h>
//#include <sys/wait.h>
//#include <stdbool.h>
//#include <string.h>
//#include <signal.h>
//
//#define clear() printf("\033[H\033[J") //clears the page
//
//const int MAX = 1000; //maximum size of arrays in program
//char** TOKEN; //list of commands
//
//void welcome(); //welcome page
//void printDir(); //prints the current directory
//void execute(char *input); //executes the given command
//void getInput(); //receives the input from user
//bool myCommands(char* input); //additional commands
//void error(); //prints an error
//void a();
//void b();
//void c();
//void d();
//void f();
//void g();
//char** splitter(const char * str, const char * delim);
//bool fileValidator(char* input);
//char* fileToText(FILE* file);
//void addHistory(char* input);
//void handleSigint(int sig);
//int  isEmpty(const char *str);
//FILE* removeEmptyLines(FILE *file);
//char* removeSpaces(char *string);
//void printFile(FILE *fptr);
//
//int main() {
//    signal(SIGINT, handleSigint);
//    welcome();
//    getInput();
//
//    printDir();
//
//
//    return 0;
//}
//
//char ** splitter(const char * str, const char * delim){
//    /* count words */
//    char * s = strdup(str);
//
//    if (strtok(s, delim) == 0)
//        /* no word */
//        return NULL;
//
//    int nw = 1;
//
//    while (strtok(NULL, delim) != 0)
//        nw += 1;
//
//    strcpy(s, str); /* restore initial string modified by strtok */
//
//    /* split */
//    char ** v = malloc((nw + 1) * sizeof(char *));
//    int i;
//
//    v[0] = strdup(strtok(s, delim));
//
//    for (i = 1; i != nw; ++i)
//        v[i] = strdup(strtok(NULL, delim));
//
//    v[i] = NULL; /* end mark */
//
//    free(s);
//
//    return v;
//}
//
//void getInput() {
//    char* input;
//    printDir();
//    input = readline("\n>>: ");
//    addHistory(input);
//    TOKEN = splitter(input," ");
//    execute(TOKEN[0]);
//}
//
//void execute(char *input) {
//
//    char* argument_list[] = {"", "-l", NULL};
//    argument_list[0] = input;
//    if(!strcmp("exit",input)){
//        printf("\nbye\n");
//        exit(0);
//    }
//
//    else{
//        pid_t child = fork();
//
//        if(child < 0) {
//            printf("error\n");
//            exit(0);
//        }
//
//        else if(child == 0) {
//            if(execvp(input,argument_list) == -1){
//                if(!myCommands(input)){
//                    error();
//                }
//            }
//            exit(0);
//        }
//        else {
//            wait(NULL);
//            getInput();
//        }
//
//    }
//
//}
//
//void welcome(){
//    clear();
//    printf("                  ****************** welcome to my linux shell ******************                   \n");
//    fflush(stdout);
//    sleep(1);
//    clear();
//}
//
//void printDir(){
//    char dir[MAX];
//    char * username;
//    getcwd(dir, sizeof(dir));
//    username = getenv("USER");
//    printf("%s >> %s\n",username,dir);
//}
//
//void error(){
//    printf("wrong command\ntry using help\n");
//    fflush(stdout);
//    sleep(1);
//    clear();
//}
//
//bool myCommands(char* input){
//    if(!strcmp(input,"exit"))
//        exit(0);
//    else if(!strcmp(input,"a")) {
//        a();
//        return true;
//    }
//    else if(!strcmp(input,"b")) {
//        b();
//        return true;
//    }
//    else if(!strcmp(input,"c")) {
//        c();
//        return true;
//    }
//    else if(!strcmp(input,"d")) {
//        d();
//        return true;
//    }
//    else if(!strcmp(input,"f")) {
//        f();
//        return true;
//    }
//    else if(!strcmp(input,"g")) {
//        g();
//        return true;
//    }
//
//    return false;
//}
//
//void g() {
//    if(fileValidator(TOKEN[1])){
//
//        FILE* file;
//        file = fopen(TOKEN[1],"r");
//        char* text = fileToText(file);
//        fclose(file);
//        int lineNumberToShow = 10;
//        int counter = 1;
//
//        for (char c = *text; c; c=*++text) {
//            printf("%c",c);
//            if(c == '\n')
//                counter++;
//            if(counter == lineNumberToShow+1)
//                break;
//        }
//
//    }
//
//    else {
//        printf("no such file\n");
//        error();
//    }
//
//    printf("********* g command *********\n");
//}
//
//void f() {
//    if(fileValidator(TOKEN[1])){
//        FILE* file;
//        int count = 0; // Line counter (result)
//        file = fopen(TOKEN[1], "r");
//
//
//        char* text = fileToText(file);
//        fclose(file);
//
//        for (char c = *text; c; c=*++text) {
//            if(c == '\n')
//                count++;
//        }
//
//        printf("%d",count+1);
//    }
//
//    else {
//        printf("no such file\n");
//        error();
//    }
//
//    printf("********* f command *********\n");
//}
//
//void d() {
//    if(fileValidator(TOKEN[1])){
//        FILE* file;
//        file = fopen(TOKEN[1],"r");
//        char* text = fileToText(file);
//
//        int i=0;
//        char lastchar;
//        for (char c = *text; c; c=*++text) {
//            int hashtag = 35;
//
//            if((c == hashtag) && (lastchar == '\n')){
//                while(c!=0) {
//                    c = *++text;
//                    lastchar = c;
//                }
//                c=*++text;
//            }
//            i++;
//            printf("%c",c);
//            lastchar = c;
//        }
//        fclose(file);
//    }
//
//    else {
//        printf("no such file\n");
//        error();
//    }
//
//    printf("********* c command *********\n");
//}
//
//void c() {
//    if(fileValidator(TOKEN[1])){
//
//        FILE* file;
//        file  = fopen(TOKEN[1], "r");
//
//        printf("\n%s\n",removeSpaces(fileToText(removeEmptyLines(file))));
//
//        fclose(file);
//    }
//
//    else {
//        printf("no such file\n");
//        error();
//    }
//    printf("********* c command *********\n");}
//
//void b() {
//    if(fileValidator(TOKEN[1])){
//        FILE* file;
//        file = fopen(TOKEN[1],"r");
//        char* text = fileToText(file);
//        fclose(file);
//
//        char result;
//        int i, len;
//        int max = -1;
//        int freq[MAX];
//        len = strlen(text);
//
//        for(i = 0; i < len; i++){
//            freq[text[i]]++;
//        }
//        for(i = 0; i < len; i++){
//            if(max < freq[text[i]]){
//                max = freq[text[i]];
//                result = text[i];
//            }
//        }
//
//        if(strcmp(&result," ")) {
//            if(strcmp(&result,"\n"))
//                printf("\\n \n");
//            else
//                printf("space\n");
//        }
//        else {
//            printf("%c\n", result);
//        }
//    }
//
//    else {
//        printf("no such file\n");
//        error();
//    }
//
//    printf("********* b command *********\n");
//}
//
//void a() {
//    if(fileValidator(TOKEN[1])){
//        FILE* file;
//        file = fopen(TOKEN[1],"r");
//        char* text = fileToText(file);
//        fclose(file);
//        char** temp = splitter(text," ");
//        puts(temp[0]);
//    }
//
//    else {
//        printf("no such file\n");
//        error();
//    }
//    printf("********* a command *********\n");
//}
//
//bool fileValidator(char* input){
//    FILE* ptr;
//    ptr = fopen(input, "r");
//    if (NULL == ptr) {
//        return false;
//    }
//    else
//        return true;
//}
//
//char* fileToText(FILE* file){
//    char ch;
//    char text[MAX];
//    int counter=0;
//
//    while (!feof(file)) {
//        ch = fgetc(file);
//        text[counter] = ch;
//        counter++;
//    }
//    fclose(file);
//
//    char* temp = malloc(MAX);
//
//    memcpy(temp,text, sizeof(text));
//    return temp;
//}
//
//void addHistory(char* input){
//    FILE* file;
//    if(fileValidator("history.txt")){
//        file = fopen("history.txt","a");
//        fprintf(file,input);
//        fprintf(file,"\n");
//        fclose(file);
//    }
//    else{
//        file = fopen("history.txt","w");
//        fprintf(file,input);
//        fprintf(file,"\n");
//        fclose(file);
//    }
//}
//
//void handleSigint(int sig){
//    printf("Caught signal %d\n", sig);//todo ^c
//}
//
//void printFile(FILE *fptr)
//{
//    char ch;
//
//    while((ch = fgetc(fptr)) != EOF)
//        putchar(ch);
//}
//
//int isEmpty(const char *str)
//{
//    char ch;
//
//    do
//    {
//        ch = *(str++);
//
//        // Check non whitespace character
//        if(ch != ' ' && ch != '\t' && ch != '\n' && ch != '\r' && ch != '\0')
//            return 0;
//
//    } while (ch != '\0');
//
//    return 1;
//}
//
//FILE* removeEmptyLines(FILE *file){
//    FILE* temp;
//    temp = fopen("remove-blanks.tmp", "w");
//
//
//    /* Exit if file not opened successfully */
//    if (file == NULL || temp == NULL)
//    {
//        printf("Unable to open file.\n");
//        printf("Please check you have read/write previleges.\n");
//
//        exit(EXIT_FAILURE);
//    }
//
//    // Move src file pointer to beginning
//    rewind(file);
//
//    // Remove empty lines from file.
//    char buffer[MAX];
//
//    while ((fgets(buffer, MAX, file)) != NULL)
//    {
//        /* If current line is not empty then write to temporary file */
//        if(!isEmpty(buffer))
//            fputs(buffer, temp);
//    }
//
//    /* Close all open files */
//    fclose(temp);
//    temp = fopen("remove-blanks.tmp", "r");
//    return temp;
//
//}
//
//char * removeSpaces(char *string)
//{
//    // non_space_count to keep the frequency of non space characters
//    int non_space_count = 0;
//
//    //Traverse a string and if it is non space character then, place it at index non_space_count
//    for (int i = 0; string[i] != '\0'; i++)
//    {
//        if (string[i] != ' ')
//        {
//            string[non_space_count] = string[i];
//            non_space_count++;//non_space_count incremented
//        }
//    }
//
//    //Finally placing final character at the string end
//    string[non_space_count] = '\0';
//    return string;
//}
//
//


main.c
#include<stdio.h>
#include<string.h>
#include<stdlib.h>
#include<unistd.h>
#include<sys/types.h>
#include<sys/wait.h>
#include<readline/readline.h>
#include<readline/history.h>
#include <sys/wait.h>
#include <stdbool.h>
#include <string.h>
#include <signal.h>

#define clear() printf("\033[H\033[J") //clears the page

const int MAX = 1000; //maximum size of arrays in program
char** TOKEN; //list of commands

void welcome(); //welcome page
void printDir(); //prints the current directory
void execute(char *input); //executes the given command
void getInput(); //receives the input from user
bool myCommands(char* input); //additional commands
void error(); //prints an error
void a(); //cuts the first string
void b(); //most frequent char in text
void c(); //deletes all empty spaces such as space,\n,\t,...
void d(); //displays all onCommented lines
void f(); //number of text lines
void g(); //first 10 lines of text or more ...
char** splitter(const char * str, const char * delim); //splits a string base on the char you want
bool fileValidator(char* input); //validates the file base of given location
char* fileToText(FILE* file); //returns a text of file
void addHistory(char* input); //adds the commands to history.txt
void handleSigint(int sig); //handles the ctrl+c
bool isEmpty(const char *str); //returns a bool based on text given to
FILE* removeEmptyLines(FILE *file); //removes empty lines of given file
char* removeSpaces(char *string); //removes space of a text
void printFile(FILE *fptr); //prints the given file as a text

int main() {
    signal(SIGINT, handleSigint);
    welcome();
    getInput();
    printDir();

    return 0;
}

char ** splitter(const char * str, const char* delim){
    /* count words */
    char * s = strdup(str);

    if (strtok(s, delim) == 0)
        /* no word */
        return NULL;

    int nw = 1;

    while (strtok(NULL, delim) != 0)
        nw += 1;

    strcpy(s, str); /* restore initial string modified by strtok */

    /* split */
    char ** v = malloc((nw + 1) * sizeof(char *));
    int i;

    v[0] = strdup(strtok(s, delim));

    for (i = 1; i != nw; ++i)
        v[i] = strdup(strtok(NULL, delim));

    v[i] = NULL; /* end mark */

    free(s);

    return v;
}

void getInput() {
    char* input;
    printDir();
    input = readline("\n>>: ");
    addHistory(input);
    TOKEN = splitter(input," ");
    execute(TOKEN[0]);
}

void execute(char *input) {

    char* argument_list[] = {"", "-l", NULL};
    argument_list[0] = input;
    if(!strcmp("exit",input)){
        printf("\nbye\n");
        exit(0);
    }

    else{
        pid_t child = fork();

        if(child < 0) {
            printf("error\n");
            exit(0);
        }

        else if(child == 0) {
            if(execvp(input,argument_list) == -1){
                if(!myCommands(input)){
                    error();
                }
            }
            exit(0);
        }
        else {
            wait(NULL);
            getInput();
        }

    }

}

void welcome(){
    clear();
    printf("                  ****************** welcome to my linux shell ******************                   \n");
    fflush(stdout);
    sleep(1);
    clear();
}

void printDir(){
    char dir[MAX];
    char * username;
    getcwd(dir, sizeof(dir));
    username = getenv("USER");
    printf("%s >> %s\n",username,dir);
}

void error(){
    fprintf(stderr,"wrong command\ntry using help\n");
    fflush(stdout);
    sleep(1);
    clear();
}

bool myCommands(char* input){
    if(!strcmp(input,"exit"))
        exit(0);
    else if(!strcmp(input,"a")) {
        a();
        return true;
    }
    else if(!strcmp(input,"b")) {
        b();
        return true;
    }
    else if(!strcmp(input,"c")) {
        c();
        return true;
    }
    else if(!strcmp(input,"d")) {
        d();
        return true;
    }
    else if(!strcmp(input,"f")) {
        f();
        return true;
    }
    else if(!strcmp(input,"g")) {
        g();
        return true;
    }

    return false;
}

void g() {
    if(fileValidator(TOKEN[1])){

        FILE* file;
        file = fopen(TOKEN[1],"r");
        char* text = fileToText(file);
        fclose(file);
        int lineNumberToShow = 10;
        int counter = 1;

        for (char c = *text; c; c=*++text) {
            printf("%c",c);
            if(c == '\n')
                counter++;
            if(counter == lineNumberToShow+1)
                break;
        }

    }

    else {
        printf("no such file\n");
        error();
    }

    printf("********* g command *********\n");
}

void f() {
    if(fileValidator(TOKEN[1])){
        FILE* file;
        int count = 0; // Line counter (result)
        file = fopen(TOKEN[1], "r");


        char* text = fileToText(file);
        fclose(file);

        for (char c = *text; c; c=*++text) {
            if(c == '\n')
                count++;
        }

        printf("%d",count+1);
    }

    else {
        printf("no such file\n");
        error();
    }

    printf("********* f command *********\n");
}

void d() {
    if(fileValidator(TOKEN[1])){
        FILE* file;
        file = fopen(TOKEN[1],"r");
        char* text = fileToText(file);

        int i=0;
        char lastchar;
        for (char c = *text; c; c=*++text) {
            int hashtag = 35;

            if((c == hashtag) && (lastchar == '\n')){
                while(c!=0) {
                    c = *++text;
                    lastchar = c;
                }
                c=*++text;
            }
            i++;
            printf("%c",c);
            lastchar = c;
        }
        fclose(file);
    }

    else {
        printf("no such file\n");
        error();
    }

    printf("********* c command *********\n");
}

void c() {
    if(fileValidator(TOKEN[1])){

        FILE* file;
        file  = fopen(TOKEN[1], "r");

        printf("\n%s\n",removeSpaces(fileToText(removeEmptyLines(file))));

        fclose(file);
    }

    else {
        printf("no such file\n");
        error();
    }
    printf("********* c command *********\n");}

void b() {
    if(fileValidator(TOKEN[1])){
        FILE* file;
        file = fopen(TOKEN[1],"r");
        char* text = fileToText(file);
        fclose(file);

        char result;
        int i, len;
        int max = -1;
        int freq[MAX];
        len = strlen(text);

        for(i = 0; i < len; i++){
            freq[text[i]]++;
        }
        for(i = 0; i < len; i++){
            if(max < freq[text[i]]){
                max = freq[text[i]];
                result = text[i];
            }
        }

        if(strcmp(&result," ")) {
            if(strcmp(&result,"\n"))
                printf("\\n \n");
            else
                printf("space\n");
        }
        else {
            printf("%c\n", result);
        }
    }

    else {
        printf("no such file\n");
        error();
    }

    printf("********* b command *********\n");
}

void a() {
    if(fileValidator(TOKEN[1])){
        FILE* file;
        file = fopen(TOKEN[1],"r");
        char* text = fileToText(file);
        fclose(file);
        char** temp = splitter(text," ");
        puts(temp[0]);
    }

    else {
        printf("no such file\n");
        error();
    }
    printf("********* a command *********\n");
}

bool fileValidator(char* input){
    FILE* ptr;
    ptr = fopen(input, "r");
    if (NULL == ptr) {
        return false;
    }
    else
        return true;
}

char* fileToText(FILE* file){
    char ch;
    char text[MAX];
    int counter=0;

    while (!feof(file)) {
        ch = fgetc(file);
        text[counter] = ch;
        counter++;
    }
    fclose(file);

    char* temp = malloc(MAX);

    memcpy(temp,text, sizeof(text));
    return temp;
}

void addHistory(char* input){
    FILE* file;
    if(fileValidator("history.txt")){
        file = fopen("history.txt","a");
        fprintf(file,input);
        fprintf(file,"\n");
        fclose(file);
    }
    else{
        file = fopen("history.txt","w");
        fprintf(file,input);
        fprintf(file,"\n");
        fclose(file);
    }
}

void handleSigint(int sig){
    clear();
    printDir();
}

void printFile(FILE *fptr)
{
    char ch;

    while((ch = fgetc(fptr)) != EOF)
        putchar(ch);
}

bool isEmpty(const char *str)
{
    char ch;

    do
    {
        ch = *(str++);

        // Check non whitespace character
        if(ch != ' ' && ch != '\t' && ch != '\n' && ch != '\r' && ch != '\0')
            return true;

    } while (ch != '\0');

    return false;
}

FILE* removeEmptyLines(FILE *file){
    FILE* temp;
    temp = fopen("remove-blanks.tmp", "w");

    rewind(file);

    char buffer[MAX];

    while ((fgets(buffer, MAX, file)) != NULL)
    {
        // If current line is not empty then write to temporary file
        if(!isEmpty(buffer))
            fputs(buffer, temp);
    }

    fclose(temp);
    temp = fopen("remove-blanks.tmp", "r");
    return temp;

}

char * removeSpaces(char *string){
    int non_space_count = 0;

    for (int i = 0; string[i] != '\0'; i++)
    {
        if (string[i] != ' ')
        {
            string[non_space_count] = string[i];
            non_space_count++;
        }
    }

    string[non_space_count] = '\0';
    return string;
}


